File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* temp = head, *res = nullptr;

        if(!head || !head->next) return res;

        int len = 0;
        while(temp){
            len++;
            temp=temp->next;
        }

        int target = len - n, ctr = 1;
        ListNode* mover = head;
        while(mover && ctr < target){
            ctr++;
            mover = mover->next;
        }

        ListNode*ptr = nullptr;
        if(target == 0){
            ptr = head;
            head = head->next;
        } else {
            ptr = mover->next;
            mover->next = ptr->next;
        }
            delete ptr;

        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let sz = 0;
    let mover = head;
    while(mover){
        sz++;
        mover = mover.next;
    }

    let pos = sz-n+1, ctr = 1, prev = null;
    mover = head;
    
    while(ctr != pos){
        ctr++;
        prev = mover;
        mover = mover.next;
    }

    if(prev) prev.next = mover.next;
    else head = head.next;
    
    return head;
};